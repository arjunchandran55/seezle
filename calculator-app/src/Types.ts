const TYPES = {
    CalculatorStore: Symbol.for("CalculatorStore")
};

export { TYPES };