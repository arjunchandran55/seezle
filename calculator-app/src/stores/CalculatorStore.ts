import { injectable } from "inversify";
import { observable, action, makeObservable } from "mobx";
import {config} from "../config/config";

@injectable()
export class CalculatorStore {
    @observable inputExpression: string;
    @observable result: string;
    private defaultExpression = "0";

    constructor() {
        makeObservable(this);
        this.inputExpression = this.defaultExpression;
        this.result = "";
    }

    @action calculate = () => {
        fetch(`${config.apiUrl}/basic-calculator?expression=${(encodeURIComponent(this.inputExpression))}`)
        .then(response => response.json())
        .then(response => {
            this.result = response.expressionResult as string;
        });
    }

    @action setExpression = (expression: string) => {
            this.inputExpression = expression;
    }

    @action clearExpression = () => {
        this.inputExpression = this.defaultExpression;
        this.result = "";
    }

    @action deleteLast = () => {
        this.inputExpression = this.inputExpression.slice(0, this.inputExpression.length - 1);
        this.result = "";
    }
}