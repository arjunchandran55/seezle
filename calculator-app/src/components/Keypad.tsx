import { useInjection } from "inversify-react";
import { Button, ButtonGroup, Container, Row } from "react-bootstrap";
import {CalculatorStore} from "../stores/CalculatorStore";
import { observer } from 'mobx-react';
import { TYPES } from "../Types";

const Keypad = () => {
    const store: CalculatorStore = useInjection(TYPES.CalculatorStore);

    const lastInputValue = () => store.inputExpression.slice((store.inputExpression.length - 1), store.inputExpression.length);

    const isOperator = (expr: string) => expr.includes("+") || expr.includes("-") || expr.includes("*") || expr.includes("/") || expr.includes(".") || expr.includes("=");

    const setExpression = (event: any) => {
        const lastInput = lastInputValue();
        const currentInput = event.target.value as string;
        
        if(isOperator(lastInput) && isOperator(currentInput)) {
            store.setExpression(store.inputExpression.substring(0, store.inputExpression.length - 1) + currentInput);
        }
        else if(store.inputExpression === "0") store.setExpression(currentInput);
        else store.setExpression(store.inputExpression + currentInput);
    };

    const calculateExpression = () => {
        store.calculate();
    }

    return <Container>
    <br/>
    <Row>
        <ButtonGroup size="lg" className="justify">
            <Button variant="primary" size="lg" className="btn-num" block onClick={() => store.clearExpression() }>CL</Button> &nbsp;
            <Button variant="primary" size="lg" className="btn-num" onClick={() => store.deleteLast() }> {"<"} </Button>
        </ButtonGroup>
    </Row> <br/>
    <Row>
      <ButtonGroup size="lg" className="mb-2">
          <Button variant="primary" size="sm" className="btn-num" value={1} onClick={setExpression}>1</Button>
          <Button variant="primary" size="sm" className="btn-num" value={2} onClick={setExpression}>2</Button>
          <Button variant="primary" size="sm" className="btn-num" value={3} onClick={setExpression}>3</Button> &nbsp;
          <Button variant="primary" size="sm" className="btn-num" value={"/"} onClick={setExpression}> / </Button>
      </ButtonGroup>
    </Row>

      <Row>
          <ButtonGroup size="lg" className="mb-2">
              <Button variant="primary" size="sm" className="btn-num" value={4} onClick={setExpression}>4</Button>
              <Button variant="primary" size="sm" className="btn-num" value={5} onClick={setExpression}>5</Button>
              <Button variant="primary" size="sm" className="btn-num" value={6} onClick={setExpression}>6</Button> &nbsp;
              <Button variant="primary" size="sm" className="btn-num" value={"+"} onClick={setExpression}>+</Button>
          </ButtonGroup>
      </Row>

      <Row>
          <ButtonGroup size="lg" className="mb-2">
              <Button variant="primary" size="sm" className="btn-num" value={7} onClick={setExpression}>7</Button>
              <Button variant="primary" size="sm" className="btn-num" value={8} onClick={setExpression}>8</Button>
              <Button variant="primary" size="sm" className="btn-num" value={9} onClick={setExpression}>9</Button> &nbsp;
              <Button variant="primary" size="sm" className="btn-num" value={"-"} onClick={setExpression}>-</Button>
          </ButtonGroup>
      </Row>

      <Row>
          <ButtonGroup size="lg" className="mb-2">
              <Button variant="primary" size="sm" className="btn-num" value={"."} onClick={setExpression}>.</Button>
              <Button variant="primary" size="sm" className="btn-num" value={0} onClick={setExpression}>0</Button>
              <Button variant="primary" size="sm" className="btn-num" value={"="} onClick={calculateExpression}>=</Button> &nbsp;
              <Button variant="primary" size="sm" className="btn-num" value={"*"} onClick={setExpression}>x</Button>
          </ButtonGroup>
      </Row>

      <ButtonGroup size="lg" className="mb-2" vertical>
      </ButtonGroup>
  </Container>
};

export default observer(Keypad);