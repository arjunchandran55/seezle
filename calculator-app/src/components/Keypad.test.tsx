import { configure, render, shallow } from "enzyme";
import Adapter from 'enzyme-adapter-react-16';
import {CalculatorStore} from "../stores/CalculatorStore";
import { Provider } from 'inversify-react';
import { Container } from 'inversify';
import { TYPES } from '../Types';
import "reflect-metadata";
import Keypad from "./Keypad";

configure({ adapter: new Adapter() }); // configure enzyme adapter

var containerMock: Container;
var storeMock: CalculatorStore;
beforeEach(() => {
  
  containerMock = new Container();
  
  storeMock = jest.genMockFromModule(
    "../stores/CalculatorStore",
  ) as CalculatorStore;

  storeMock.setExpression = jest.fn(e => { storeMock.inputExpression = e } );
  containerMock.bind(TYPES.CalculatorStore).toConstantValue(storeMock);
});

const renderComponent = (renderer: any) =>
  renderer(
    <Provider container={containerMock}>
      <Keypad />
    </Provider>
  );

test("Render Keypad for calculator", () => {
  const comp = renderComponent(shallow);
  expect(comp).toMatchSnapshot();
});