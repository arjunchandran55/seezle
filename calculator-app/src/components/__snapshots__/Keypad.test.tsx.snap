// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Keypad for calculator 1`] = `ShallowWrapper {}`;
